try.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022, Joe Tsai. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE.md file.
+
+// Package try emulates aspects of the ill-fated "try" proposal using generics.
+// See https://golang.org/issue/32437 for inspiration.
+//
+// Example usage:
+//
+//	func Fizz(...) (..., err error) {
+//		defer try.Catch(&err, func() {
+//			if err == io.EOF {
+//				err = io.ErrUnexpectedEOF
+//			}
+//		})
+//		... := try.E2(Buzz(...))
+//		return ..., nil
+//	}
+//
+// This package is not intended for production critical code as quick and easy
+// error handling can occlude critical error handling logic.
+// Rather, it is intended for short Go programs and unit tests where
+// development speed is a greater priority than reliability.
+//
+//
+// Code before try:
+//
+//	func (a *MixedArray) UnmarshalNext(uo json.UnmarshalOptions, d *json.Decoder) error {
+//		switch t, err := d.ReadToken(); {
+//		case err != nil:
+//			return err
+//		case t.Kind() != '[':
+//			return fmt.Errorf("got %v, expecting array start", t.Kind())
+//		}
+//
+//		if err := uo.UnmarshalNext(d, &a.Scalar); err != nil {
+//			return err
+//		}
+//		if err := uo.UnmarshalNext(d, &a.Slice); err != nil {
+//			return err
+//		}
+//		if err := uo.UnmarshalNext(d, &a.Map); err != nil {
+//			return err
+//		}
+//
+//		switch t, err := d.ReadToken(); {
+//		case err != nil:
+//			return err
+//		case t.Kind() != ']':
+//			return fmt.Errorf("got %v, expecting array start", t.Kind())
+//		}
+//		return nil
+//	}
+//
+// Code after try:
+//
+//	func (a *MixedArray) UnmarshalNext(uo json.UnmarshalOptions, d *json.Decoder) (err error) {
+//		defer try.Catch(&err)
+//		if t := try.E1(d.ReadToken()); t.Kind() != '[' {
+//			return fmt.Errorf("found %v, expecting array start", t.Kind())
+//		}
+//		try.E(uo.UnmarshalNext(d, &a.Scalar))
+//		try.E(uo.UnmarshalNext(d, &a.Slice))
+//		try.E(uo.UnmarshalNext(d, &a.Map))
+//		if t := try.E1(d.ReadToken()); t.Kind() != ']' {
+//			return fmt.Errorf("found %v, expecting array start", t.Kind())
+//		}
+//		return nil
+//	}
+//
+package try
+
+// wrapError wraps an error to ensure that we only recover from errors
+// panicked by this package.
+type wrapError struct{ error }
+
+// Unwrap primarily exists for testing purposes.
+func (e wrapError) Unwrap() error { return e.error }
+
+// Catch catches a previously panicked error and stores it into err.
+// If it successfully catches an error, it calls any provided handlers.
+func Catch(err *error, handlers ...func()) {
+	switch ex := recover().(type) {
+	case nil:
+		return
+	case wrapError:
+		*err = ex.error
+		for _, handler := range handlers {
+			handler()
+		}
+	default:
+		panic(ex)
+	}
+}
+
+// E panics if err is non-nil.
+func E(err error) {
+	if err != nil {
+		panic(wrapError{err})
+	}
+	return
+}
+
+// E1 panics if err is non-nil,
```

Comment:
I'd phrase these a bit differently.

```go
// E1 returns v1. It panics if err is non-nil.
```

This emphasizes the happy path, obviates the need for "otherwise" and "as is", and gets it all on a single line.

Reply:
Done.